pyomo/contrib/cspline_external/cspline_parameters.py

```diff
@@ -0,0 +1,244 @@
+
+    Returns:
+        Poymo ConcreteModel
+    """
+    n_data = len(x_data)
+    assert n_data == len(y_data)
+    if x_knots is None:
+        n_knots = n_data
+        x_knots = x_data
+    else:
+        n_knots = len(x_knots)
+
+    m = pyo.ConcreteModel(name=name)
+    # Sets of indexes for knots, segments, and data
+    m.knt_idx = pyo.RangeSet(n_knots)
+    m.seg_idx = pyo.RangeSet(n_knots - 1)
+    m.dat_idx = pyo.RangeSet(n_data)
+
+    m.x_data = pyo.Param(m.dat_idx, initialize={i + 1: x for i, x in enumerate(x_data)})
+    m.y_data = pyo.Param(m.dat_idx, initialize={i + 1: x for i, x in enumerate(y_data)})
+    m.x = pyo.Param(m.knt_idx, initialize={i + 1: x for i, x in enumerate(x_knots)})
+    m.alpha = pyo.Var(m.seg_idx, {1, 2, 3, 4}, initialize=1)
+
+    # f_s(x) = f_s+1(x)
+    @m.Constraint(m.seg_idx)
+    def y_eqn(blk, s):
+        if s == m.seg_idx.last():
+            return pyo.Constraint.Skip
+        return _f_cubic(m.x[s + 1], m.alpha, s) == _f_cubic(m.x[s + 1], m.alpha, s + 1)
+
+    # f'_s(x) = f'_s+1(x)
+    @m.Constraint(m.seg_idx)
+    def yx_eqn(blk, s):
+        if s == m.seg_idx.last():
+            return pyo.Constraint.Skip
+        return _fx_cubic(m.x[s + 1], m.alpha, s) == _fx_cubic(
+            m.x[s + 1], m.alpha, s + 1
+        )
+
+    # f"_s(x) = f"_s+1(x)
+    @m.Constraint(m.seg_idx)
+    def yxx_eqn(blk, s):
+        if s == m.seg_idx.last():
+            return pyo.Constraint.Skip
+        return _fxx_cubic(m.x[s + 1], m.alpha, s) == _fxx_cubic(
+            m.x[s + 1], m.alpha, s + 1
+        )
+
+    # Identify segments used to predict y_data at each x_data.  We use search in
+    # instead of a dict lookup, since we don't want to require the data to be at
+    # the knots, even though that is almost always the case.
+    idx = searchsorted(x_knots, x_data)
+
+    if end_point_constraint:
+        add_endpoint_second_derivative_constraints(m)
+
+    # Expression for difference between data and prediction
+    @m.Expression(m.dat_idx)
+    def ydiff(blk, d):
+        s = idx[d - 1] + 1
+        if s >= m.seg_idx.last():
+            s -= 1
+        return m.y_data[d] - _f_cubic(m.x_data[d], m.alpha, s)
+
+    if objective_form:
+        # least squares objective
+        m.obj = pyo.Objective(expr=sum(m.ydiff[d] ** 2 for d in m.dat_idx))
+    else:
+
+        @m.Constraint(m.dat_idx)
+        def match_data(blk, d):
+            return m.ydiff[d] == 0
+
+    return m
+
+
+def add_endpoint_second_derivative_constraints(m):
+    """Usually cubic splines use the endpoint constraints that the second
+    derivative is zero.  This function adds those constraints to a model
+    """
+
+    @m.Constraint([m.seg_idx.first(), m.seg_idx.last()])
+    def yxx_endpoint_eqn(blk, s):
+        if s == m.seg_idx.last():
+            j = m.knt_idx.last()
+        else:
+            j = s
+        return _fxx_cubic(m.x[j], m.alpha, s) == 0
+
+
+def get_parameters(m, file_name=None):
```

Comment:
I think I would prefer to separate this into two functions: `get_parameters()` and `write_parameters()`: when reading the current code, it looks like the parameters are being *read* from the provided file.

Additional thought: Should get_parameters return a data structure (or at least a namedtuple?)

Additional thought: it might be better / easier (to test) if `write_parameters` took the parameters data structure and a file-like object instead of a filename (that will both be more consistent witht eh newer Pyomo way of doing things, and allow writing the params to other locations (like StringIO, loggers, pipes, etc).

Reply:
Good point.  Using this recently I was confused by this for a bit.

Reply:
Okay, I reworked this in a way that will hopefully addresses all this.